content_script.js

```diff
@@ -1,82 +1,213 @@
+        for (var i = 0, child; child = children[i]; i++) {
+          this.serializeTree(child, depth+1);
+        }
+      }
+
+      this.html.push(new Array(depth+1).join('  '));
+      this.html.push(`</${element.tagName.toLowerCase()}>\n`);
     }
-    
-    var children = element.childNodes;
-    if (children) {
-      for (var i = 0; i < children.length; i++) {
-        var child = children[i];
-        html.push(serializeTree(html, holes, child, depth+1));
+  }
+
+  /**
+   * Takes an html document, and populates this objects fields such that it can eventually be converted into an
+   * html file.
+   *
+   * @param {Document} doc The Document to serialize. Defaults to the current document
+   * @public
+   */ 
+  serializeDocument(doc) {
+    doc = doc || document;
+    this.html.push('<!DOCTYPE html>\n');
+    var node = doc.firstChild;
+    while (node.nodeType == Node.DOCUMENT_TYPE_NODE) {
+      node = node.nextSibling; // TODO: could this cause an infinite loop?
+    }
+    this.serializeTree(node, 0);
+  }
+
+  /**
+   * Computes the index of the window in it's parent's array of frames.
+   *
+   * @param {Window} win The window to use in the calculation
+   * @return {number} the frames index
+   */
+  iframePosition(win) {
+    if (win.parent != win) {
+      for (var i = 0; i < win.parent.frames.length; i++) {
+        if (win.parent.frames[i] == win) {
+          return i;
+        }
       }
+    } else {
+      return -1;
     }
+  }
 
-    html.push(new Array(depth+1).join("  "));
-    html.push("</" + element.tagName + ">\n");
+  /**
+   * Computes the full path of the frame in the root document.  Nested layers are seperated by '.'
+   *
+   * @param {Window} win The window to use in the calculation
+   * @return {string} The full path
+   */
+  iframeFullPosition(win) {
+    if (this.iframePosition(win) < 0) {
+      return '0';
+    } else {
+      return this.iframeFullPosition(win.parent) + '.' + this.iframePosition(win); 
+    }
   }
 }
+/////////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////////
 
-function fillHoles(html, holes) {
-  if (Object.keys(holes).length == 0) {
-    var output = "<!DOCTYPE html>\n" + html.join("");
-    var file = new Blob([output], {type: 'text/html'});
-    var url = URL.createObjectURL(file)
-    // var extensionId = "nlhglfcihneopknjknodakgjfiapnnmn";
-    chrome.runtime.sendMessage(url);
-  }
-  else {
-    var index = Object.keys(holes)[0];
-    var src = holes[index];
-    delete holes[index];
+/**
+ * Takes all of the srcHoles in the HTMLSerializer starting at index i, and creates data
+ * urls for the resources, and places them in this.html
+ *
+ * @param {HTMLSerializer} htmlSerializer The HTMLSerializer
+ * @param {number} i The index of this.srcHoles at which to start
+ * @param {Function} callback The callback function
+ */
+function fillSrcHoles(htmlSerializer, i, callback) {
```

Comment:
This should be an internal helper method. The second argument is an implementation detail:

function fillSrcHoles(htmlSerializer, callback) {
  fillSrcHolesInternal(htmlSerializer, 0,  callback)
}

function fillSrcHoles(htmlSerializer, 0, callback) {
 ...
}